internal/scheduler/scheduler.go

```diff
@@ -0,0 +1,293 @@
+	const op = "scheduler.(Scheduler).UpdateJobNextRun"
+	if name == "" {
+		return errors.New(errors.InvalidParameter, op, "missing name")
+	}
+	repo, err := s.jobRepoFn()
+	if err != nil {
+		return errors.Wrap(err, op)
+	}
+
+	_, err = repo.UpdateJobNextRun(ctx, name, nextRunIn)
+	if err != nil {
+		return errors.Wrap(err, op)
+	}
+	return nil
+}
+
+// Start begins the scheduling loop that will query the repository for jobs to run and
+// run them in a goroutine, the scheduler will stop all running jobs and stop requesting
+// new jobs once the ctx past in is cancelled.
+// The scheduler cannot be started again once the ctx is cancelled, a new scheduler will
+// need to be instantiated in order to begin scheduling again.
+func (s *Scheduler) Start(ctx context.Context) error {
+	const op = "scheduler.(Scheduler).Start"
+	if !s.started.CAS(s.started.Load(), true) {
+		s.logger.Debug("scheduler already started, skipping")
+		return nil
+	}
+
+	if err := ctx.Err(); err != nil {
+		return err
+	}
+
+	repo, err := s.jobRepoFn()
+	if err != nil {
+		return errors.Wrap(err, op)
+	}
+
+	// Interrupt all runs previously allocated to this server
+	_, err = repo.InterruptRuns(ctx, 0, job.WithServerId(s.serverId))
+	if err != nil {
+		return errors.Wrap(err, op)
+	}
+
+	go s.start(ctx)
+	go s.monitorJobs(ctx)
+
+	return nil
+}
+
+func (s *Scheduler) start(ctx context.Context) {
+	s.logger.Debug("starting scheduling loop")
+	timer := time.NewTimer(0)
+	for {
+		select {
+		case <-ctx.Done():
+			s.logger.Debug("scheduling loop shutting down")
+			return
+		case <-timer.C:
+			s.logger.Debug("waking up to run jobs")
+
+			repo, err := s.jobRepoFn()
+			if err != nil {
+				s.logger.Error("error creating job repo", "error", err)
+				break
+			}
+
+			runs, err := repo.RunJobs(ctx, s.serverId, job.WithRunJobsLimit(s.runJobsLimit))
+			if err != nil {
+				s.logger.Error("error getting jobs to run from repo", "error", err)
+				break
+			}
+
+			for _, r := range runs {
+				err := s.runJob(ctx, r)
+				if err != nil {
+					s.logger.Error("error starting job", "error", err)
+					if _, inner := repo.FailRun(ctx, r.PrivateId); inner != nil {
+						s.logger.Error("error updating failed job run", "error", inner)
+					}
+				}
+			}
+		}
+
+		s.logger.Debug("scheduling loop going back to sleep")
+		timer.Reset(s.runJobsInterval)
+	}
+}
+
+func (s *Scheduler) runJob(ctx context.Context, r *job.Run) error {
+	regJob, ok := s.registeredJobs.Load(r.JobName)
+	if !ok {
+		return fmt.Errorf("job %q not registered on scheduler", r.JobName)
+	}
+	s.logger.Debug("starting job run", "run id", r.PrivateId, "job name", r.JobName)
+
+	repo, err := s.jobRepoFn()
+	if err != nil {
+		return fmt.Errorf("failed to create job repository: %w", err)
+	}
+
+	j := regJob.(Job)
+	jobContext, jobCancel := context.WithCancel(ctx)
+	_, loaded := s.runningJobs.LoadOrStore(r.JobName, runningJob{runId: r.PrivateId, cancelCtx: jobCancel, status: j.Status})
```

Comment:
Is there a reason you're not storing a pointer to the runningJob struct? It doesn't strictly need to be one, but you could then modify the values after the fact if you did so. For instance, you could create the context only if loaded, so not have to ensure that it's canceled in all subsequent paths.